libcob/cconv.c

```diff
@@ -471,3 +471,84 @@ cob_get_collation_by_name (const char *col_name,
 
 	return col_id;
 }
+
+static int
+cob_convert_hex_digit(char h)
+{
+	if (h >= '0' && h <= '9') return h - '0';
+	else if (h >= 'A' && h <= 'F') return 10 + h - 'A';
+	else if (h >= 'a' && h <= 'f') return 10 + h - 'a';
+	else return -1;
+}
+
+static int
+cob_convert_hex_byte(const char *h)
+{
+	int d1 = cob_convert_hex_digit(h[0]);
+	int d2 = cob_convert_hex_digit(h[1]);
+	if (d1 < 0 || d2 < 0) {
+		return -1;
+	} else {
+		return d1 * 16 + d2;
+	}
+}
+
+int
+cob_load_collation (const char *col_name,
+		    cob_u8_t *ebcdic_to_ascii,
+		    cob_u8_t *ascii_to_ebcdic)
+{
+	char hex[512];
+	int n, i;
+	FILE *f;
+
+	f = fopen(col_name, "rb");
+	if (f == NULL) {
+		cob_runtime_error (_("Can't open translation table '%s'"), col_name);
+		return -1;
+	}
+
+	n = fread(hex, 1, 512, f);
+	if (n != 512) {
+		if (ferror(f)) {
+			cob_runtime_error (_("Error reading translation table '%s'"), col_name);
+		} else if (feof(f)) {
+			cob_runtime_error (_("Not enough data in translation table '%s'"), col_name);
+		}
+		fclose(f);
+		return -1;
+	}
+
+	/* Allow trailing whitespaces */
+	while (!feof(f)) {
+		switch (fgetc(f)) {
+		case -1:
+		case ' ': case '\t': case '\r':	case '\n':
+			break;
+		default:
+			cob_runtime_error (_("Too much data in translation table '%s'"), col_name);
+			fclose(f);
+			return -1;
+		}
+	}
+
+	fclose(f);
+
+	for (i = 0; i < 256; ++i) {
+		n = cob_convert_hex_byte(&hex[2 * i]);
+		if (n < 0) {
+			cob_runtime_error (_("Invalid hex byte in translation table '%s' at index %d: '%c%c'"), col_name, i, hex[2 * i], hex[2 * i + 1]);
```

Comment:
Not a request, mainly wondering:
Should we do this like for the configuration?

```c
if (col_name != last_col_name) {
   cob_runtime_error (_("Errors in translation table '%s':"), col_name);
   last_col_name = col_name;
}
```
-> one message with the filename, then N detail messages without the filename?